hrjobcontract/CRM/Hrjobcontract/Upgrader.php

```diff
@@ -1065,41 +1065,91 @@ public function createInsurancePlanTypes() {
   }
 
   /**
-   * Changes Pension Provider to be a contact, instead of an option_value.
+   * Changes Pension Provider to be a contact, instead of an option_value.  All
+   * required changes are wrapped in a transaction and rolled-back if any 
+   * problems arise.
    * 
-   * @return true
+   * @return boolean
+   *   true on success
+   * 
+   * @throws Exception
    */
   function upgrade_1026() {
     $messages = [];
     $pensionsTableName = CRM_Hrjobcontract_BAO_HRJobPension::getTableName();
 
-    $this->createPensionProviderType($messages);
-    $this->replacePensionProviders($pensionsTableName, $messages);
-    $this->removePensionTypeOptionGroup($messages);
-
-    CRM_Core_DAO::executeQuery("
-      ALTER TABLE {$pensionsTableName} 
-      MODIFY pension_type INT(10) unsigned,
-      ADD CONSTRAINT pension_type_contact_id_fk
-      FOREIGN KEY(pension_type)
-      REFERENCES civicrm_contact(id)
-      ON DELETE SET NULL ON UPDATE CASCADE;
-    ");
+    $tx = new CRM_Core_Transaction();
+    try {
+      $this->createPensionProviderType($messages);
+      $this->replacePensionProviders($pensionsTableName, $messages);
+      $this->removePensionTypeOptionGroup($messages);
+      $this->alterPensionsTable($pensionsTableName, $messages);
+    } catch (Exception $ex) {
+      $tx->rollback();
+      throw new Exception('Cannot do upgrade - ' . $ex->getMessage());
+    }
 
     if (!empty($messages) && function_exists('drupal_set_message')) {
```

Comment:
I think this will never get executed. If $messages is not empty, it means some of the methods have thrown an exception, right? And if that is the case, then the block inside the catch will be executed. This block also throws an Exception and the method flow will stop there, never reaching this if